### DISCUSSION TITLE
2024-06-20 - ✨ Architecture Decision Tree - Chewing the Fat 📉

### DISCUSSION THREAD

**Discussion by sethdaily** [SSWConsulting/SSW.Rules.Content#8761] [Chewing the Fat]
Hey architects,

1. What do you think about the decision tree?
2. Do you think it would be helpful when starting a new project?

Rule: https://www.ssw.com.au/rules/software-architecture-decision-tree/

**Comment by brookjeynes-ssw** [comment]
> What do you think about the decision tree?

I think it's great. I fear it may be taken as the be-all-end-all by some but as long as it's just used as a guideline, it's a great tool to 

> Do you think it would be helpful when starting a new project?

Definitely!

**Comment by matt-goldman-ssw** [comment]
> What do you think about the decision tree?

I think it's fantastic. I think this video + rule is the single most valuable piece of public IP SSW has created since I've worked here. We should absolutely send this live ASAP and promote the heck out of it.

Equally, this is probably the most valuable CTF I can remember too.

This is something that provides genuine value - actual, tangible, real-world guidance that developer can actually, really use in actual, real projects.  We need to get this out there and do more of this sort of thing.

> Do you think it would be helpful when starting a new project?

Understatement, I think it's invaluable.

**Reply by danielmackay** [reply]
Awesome feedback!  Thanks @matt-goldman-ssw. 🙂

**Comment by AlexRothwell** [comment]
> What do you think about the decision tree?

I think it's content is great. In its current form I find it a little hard to follow with small text and yes/no not having a consistent side but these are relatively minor points

> Do you think it would be helpful when starting a new project?

Yes, absolutely. It could also easily be part of the spec review process

**Comment by ozdemir-mehmet** [comment]
What do you think about the decision tree?

It's more like a decision pyramid, but I thought it was easy to follow

Do you think it would be helpful when starting a new project?

Maybe...

**Comment by jakebayliss** [comment]
What do you think about the decision tree?
Awesome, awesome, awesome. This is a great idea. I am not 100% about the 'Is your team experienced' question though, seems like it doesnt fit with the other q's and the pros and cons of vertical vs clean arc should be considered for the specific project rather than whos in the team

**Comment by bettybondoc** [comment]
Great video and rule! It’s easy to follow, even for someone who finds architecture talks very intimidating. 🙋🏻‍♀️

**Comment by MarcusIrmscherSSW** [comment]
Seems like this would be a great tool, invaluable for any new project.

**Comment by ricksu978** [comment]
> What do you think about the decision tree?

It's great to help the teams making wise decisions on software architecture.

> Do you think it would be helpful when starting a new project?

Yes, especially for new dev team.

**Comment by GabbyWard** [comment]
First off, do like that we have some questions to prompt response, so +1.

> What do you think about the decision tree?

I think it's great, it's clear and doesn't go too hard on every question that could be asked about every project, just some general questions to help guide you to an architecture you might not have thought of using for the project 😀

> Do you think it would be helpful when starting a new project?

I think it'll be helpful when I am on a larger project than I have been currently getting 😅

**Comment by TobyChurches** [comment]
> What do you think about the decision tree?

I really like this! its a super easy to follow structure that could be used by anyone.

> Do you think it would be helpful when starting a new project?

Most definitely!

**Comment by arig2024** [comment]
Seems logical. However these trees should be definitive and there are some subjective questions in the tree. eg. 'Do you have many teams?' Is many 2 or 7? Or 'Is the team experienced?' Is epxerienced 3 years or 10 years under belt, or is does this refer to having completed 3 or 30 similar projects?

**Comment by Tanya-Leahy** [comment]
What do you think about the decision tree?

It looks simple to follow and would be a great tool to help teams with deciding on what architecture to use.

Do you think it would be helpful when starting a new project?

Yes.

**Comment by calumjs** [comment]
If you follow the tree, clean architecture will now only be selected if you have an inexperienced team, and at SSW I think we never have a team without any experienced devs, at least for a serious project...so it seems like this is effectively zz-ing clean architecture - is that the desired outcome? I'd make that more explicit if so.

**Reply by danielmackay** [reply]
@calumjs - during the architecture retreat we debated a lot about CA vs VSA.  My current thoughts is that although CA is great, VSA can be more productive and is easier to migrate to a Modular Monolith as the project grows as the concepts are similar (i.e. group by feature -> group by module).  In comparison, converting a CA project to MoMo would be more difficult as you're using to thinking about layers, no features.

**Reply by calumjs** [reply]
@danielmackay that's cool, and I am not necessarily against it - but the CA template was the go-to way to begin a new project at SSW since I started, so I think changing the recommendation should be made explicit with reasoning of why the recommendation has changed (rather than implicitly by just routing most circumstances away from it on the decision tree)

**Reply by danielmackay** [reply]
Although, these are my current thoughts I don't think CA is going away anytime soon.

We still need to get more experience with VSA and try it out on a bunch of projects.

We also need to get the VSA template to the same maturity of the CA template.

**Comment by danielmackay** [comment]
> 1. What do you think about the decision tree?

I think it's awesome! 😉

> 2. Do you think it would be helpful when starting a new project?

These sorts of things will never be perfect and cover 100% of all scenarios, but I believe it provides some good general guidance.

**Comment by bahjatalaadelssw** [comment]
> What do you think about the decision tree?

It's a quick way to get a baseline approach

> Do you think it would be helpful when starting a new project?

Yes, is a good tool for juniors

**Comment by ncn-ssw** [comment]
> 1. What do you think about the decision tree?

I think it is excellent - it provides clear guidance.

> Do you think it would be helpful when starting a new project?

Yes - I expect that I will be referring to it in future.

**Comment by uly1** [comment]
LOVED IT! Demystefied the whole process and gave resources for how to biuld with 2 of them! Well done :)

**Comment by joshbermanssw** [comment]
> What do you think about the decision tree?

Its clear and streamlines questions into answers 

> Do you think it would be helpful when starting a new project

Yes, especially for someone (like me) with little experience in software architecture

**Comment by leahy268** [comment]
> What do you think about the decision tree?
Looks pretty clear to me. 

>Do you think it would be helpful when starting a new project
I'm sure it would but I'm not a Dev so I don't think I could follow it myself.

**Comment by Calinator444** [comment]
> 1. What do you think about the decision tree?

I think it's great. There's some complexity in the decision-making process. For instance during the retreat @danielmackay raised the point that you could use architecture tests in a VSA context to ensure junior devs aren't referencing code where they shouldn't be (e.g. performing domain logic in the application layer).

> 2. Do you think it would be helpful when starting a new project?

Yes. Although I'd want a second opinion from a senior on decisions like these. Choosing the right architecture is a decision with a lot of nuance. Being consultants we have to work across a large number of domains which may be more susceptible to change than we realize. A project which might initially be a good candidate for Clean Architecture might end up being better suited for VSA as requirements change.

**Comment by JackDevAU** [comment]
> 1. What do you think about the decision tree?

Awesome! Such a simple way to knowledge share something that can take years to really know well. 
I think however, even with this a senior should still be making the final call. Awesome stuff guys! 

> 2. Do you think it would be helpful when starting a new project?

100%

**Comment by duanxinhuan** [comment]
Good work!

**Comment by YalalCakirSSW** [comment]
> What do you think about the decision tree?

It's great. Easy to understand and follow. 

> Do you think it would be helpful when starting a new project?

Definitely.

**Comment by awaltos** [comment]
Great video. I enjoyed the high level thinking about which arch to go with and more importantly the reasoning behind each step. I think it uses good questions to find out more about the client's needs too. I think this would be most useful for large projects.

**Comment by steven0x51** [comment]
1. What do you think about the decision tree?
It's awsome, simplifies the deicion which architecture to choose
1. Do you think it would be helpful when starting a new project?
It's very beneficial for project initiating

**Comment by Ben0189** [comment]
1. What do you think about the decision tree?
One of the most valuable CTF👍, the decision tree is easy to follow for both technical and non technical person, it summarize of how to choose a right architecture for starting a project

2. Do you think it would be helpful when starting a new project?
Definitely, the decision tree and template are both very helpful 😊

**Comment by cvionnet** [comment]
>What do you think about the decision tree?

This decision tree is a must-have for every team and is an invaluable addition to SSW' templates! 🚀

>Do you think it would be helpful when starting a new project?

100%

**Comment by zacharykeeping** [comment]
> What do you think about the decision tree?

Love it - definitely simplifies the initial process of deciding on an architecture.

> Do you think it would be helpful when starting a new project?

Definitely

**Comment by chrisschultzssw** [comment]
> What do you think about the decision tree?

Seems easy to follow. Could it have some colour (and less jpg)?

> Do you think it would be helpful when starting a new project?

If I were a dev, yes definitely!

**Comment by BenDunkerleySSW** [comment]
1. What do you think about the decision tree?
Anything that makes decision making easier and improves transparency is an invaluable tool. Although it obviously won't make decisions for you, using this tree as a starting point is handy. 

     2. Do you think it would be helpful when starting a new project?
 Same as the above, I do think its highly valuable though very few projects will fit perfectly into the very binary "yes/no" format of the tree. Its up to the expert discretion of the team to decide the best path, using this tree to send them generally in the right direction.

**Comment by ozair-ashfaque** [comment]
> 1. What do you think about the decision tree?

It is easy to follow.

> 2. Do you think it would be helpful when starting a new project?

Yes

**Comment by AttackOnMorty** [comment]
> What do you think about the decision tree?

It's awesome. This idea can also be applied to choose other things, e.g. frameworks.

> Do you think it would be helpful when starting a new project?

Yes

**Comment by jeoffreyfischer** [comment]
Love the decision tree - makes it easier to understand each Architecture use.
I also enjoyed the structure of the video, the examples ate the end are really helpful

**Comment by jimmy1992abc** [comment]
> 1. What do you think about the decision tree?

Nobel Prize level masterpiece

> 2.Do you think it would be helpful when starting a new project?

Definitely yes - that is the difference between knowing how to build Burj Khalifa vs playing a lego

**Comment by christoment** [comment]
> What do you think about the decision tree?

This is a very solid guideline, especially in a spec review when trying to set which architecture to recommend.
Keep in mind that this is a guideline - human should still be part of the process to decide what needs to be done.

> Do you think it would be helpful when starting a new project?

Yes!

**Comment by SamBWagner** [comment]
> What do you think about the decision tree?

I think it's a good starting point. But, I doubt that it could be more than that. I'm not a huge fan of lumping microservices into it as, while I understand that Modular Monolith and Microservices are 'competitors'. They're just not the same thing. 

Microservices is a system's architecture pattern. While MoMo is a code architecture pattern. One could do MoMo in a microservice. You couldn't do a Microservice in a MoMo.

> Do you think it would be helpful when starting a new project?

Possibly. I think understanding:

1. Cost constraints
2. Resourcing
3. Domain understanding
4. Requirements

Are far more important than spending time thinking about architecture. A large sum of these decisions are just pre-made for you when you spend time understanding those 4 things.

**Comment by PennyWalker** [comment]
I think this is a great kick-off point!

Yes - It should really help guide the team when a decision is more complex and shows some of the things to be thinking about before you get started!

**Comment by AntPolkanov** [comment]
> What do you think about the decision tree?

I like it 👍
Just think that the _"Is the team experienced?"_ question should be higher up the tree. VSA and CA are easy architectures; MMA and MA are the complex ones.

> Do you think it would be helpful when starting a new project?

For sure!

**Comment by Dhruv-0987** [comment]
What do you think about the decision tree?

This might have saved us hours of headache and discussion about what arch is best, specially for me who is still learning all the things that are involved this is an amazing help.

Do you think it would be helpful when starting a new project?

Of course, this has made it so much easier

**Comment by miichaelsmedley** [comment]
Very clearly communicated. I learned a lot.

**Comment by Aibono1225** [comment]
> What do you think about the decision tree?

Looks awesome

> Do you think it would be helpful when starting a new project?

100% for sure!

**Comment by KristenHu** [comment]
> What do you think about the decision tree?

Great, it can help me a lot to make a decision definitely

> Do you think it would be helpful when starting a new project?

Yes, sure!

**Comment by AnkyChe** [comment]
> 1. What do you think about the decision tree?
Really nice and useful, love it!

> 2. Do you think it would be helpful when starting a new project?
Yes!

**Comment by jimmidier** [comment]
> What do you think about the decision tree?

Super helpful, appreciate the effort!

> Do you think it would be helpful when starting a new project?

Yes

**Comment by jerryluossw** [comment]
> What do you think about the decision tree?

I think it's great.
> Do you think it would be helpful when starting a new project?

Sure, it's really helpful.

**Comment by babakamyljanovssw** [comment]
> What do you think about the decision tree?

Informative guide and easy to follow
> Do you think it would be helpful when starting a new project?

Definitely, I will be referring to Architecture Decision Tree

**Comment by stephanfakossw** [comment]
Looks like a great tool

**Comment by amankumarrr** [comment]
> Do you think it would be helpful when starting a new project?

Yes, it would be helpful when starting the project from the beginning. Selecting the right architecture is crucial because it's the foundation of your project. If you don't choose it correctly, the project may not be sustainable in the long term.

**Comment by calumjs** [comment]
I just realised this is the first CTF topic with a descending chart emoji rather than the regular ascending chart emoji - what does that mean 🤔
![image](https://github.com/SSWConsulting/SSW.Rules.Content/assets/13717390/5fac5f6b-02ab-4c64-b8ce-6938ef337c7a)

**Reply by sethdaily** [reply]
![image](https://github.com/SSWConsulting/SSW.Rules.Content/assets/130118701/16b9b5b1-4f54-4922-8cef-2640daced427)

**Comment by MattParkerDev** [comment]
I think this will be very helpful for teams deciding which architecture to use on a new project.

**Comment by tino-liu** [comment]
>What do you think about the decision tree?
It makes a lot of sense! 
>Do you think it would be helpful when starting a new project?
For those with less experience, I think it will be very helpful.

**Comment by drwharris** [comment]
While I dont specifically agree with the decisioning, eg uncertain requirements doesnt make for a good fit for VSA (It doesnt ake for a good fit for anything really) I think its a great starting point

**Comment by isaaclombardssw** [comment]
> What do you think about the decision tree?

It's a clear and well thought out guide on a complex decision. Very valuable.

> Do you think it would be helpful when starting a new project?

Definitely.

**Comment by tkapa** [comment]
It's a good decision tree. Though I think the best architecture is whatever gets the team off the ground the fastest in greenfields projects.

One question that maybe keeps coming up - These architectures are almost always .NET focused, how can we apply them to other techs?

**Comment by vladislav-kir** [comment]
> What do you think about the decision tree?

I think it is great starting point when making architectural decision on a new project.

> Do you think it would be helpful when starting a new project?

Definitely helpful.

**Comment by JeanThirion** [comment]
I've never done VSA at scale; so I don't have much of an opinion on VSA vs CA.

I do have a lot more concerns about VSA than I have on CA though:
- Sounds messier when changing a whole horizontal slice of the App (API, UI, ORM, whatever) since it'll touch virtually all files
- Depencies will by nature be messier, since there's weaker constraints
- Keeping stuff DRY means breaking the pattern

These are the top 3 coming to mind in 5 seconds - but again, I don't know enough to judge

**Comment by Aibono1225** [comment]
Good architecture is essential for scalability, maintainability, and productivity. Getting it right early helps avoid major risks and costly rework later.

**Comment by bradystroud** [comment]
IMO Picking an architecture and sticking to it is more important than picking the perfect one.

**Reply by joshbermanssw** [reply]
+1 on this, familiarity is great

**Comment by RobJT-wq** [comment]
I'm not a dev so I don't have any war stories but I understand that choosing the right architecture for a project is important as building code from one large file would be frustrating to fix.

**Comment by StefStar** [comment]
Not a dev but if you don't "map" out what you're going to be building, regardless of what it is, you're going to have a bad time.

**Comment by TobyChurches** [comment]
> why it's important to get your Architecture right from the start of a project? Any war stories?

Getting the architecture right from the start of a project makes a massive difference in terms of scalability and maintainability. In a legacy project I’ve worked on, it became clear over time that the architecture had deteriorated as the codebase grew and multiple developers cycled on and off the team. This led to increased complexity, tight coupling, and mounting technical debt.

We've since steered the product toward a Modular Monolith architecture, and it has made a world of difference.

**Comment by ncn-ssw** [comment]
The [SugarLearning](https://sugarlearning.com/) .NET 8 upgrade was made easier by the project using Clean Architecture, as most of the rework could be isolated in the Presentation layer.

**Comment by Lewkans** [comment]
Not knowing which Architecture you are using can make it very messy as there will be no consistency in the code. It's also worth knowing the scale of your project to pick the right one before you start.

**Comment by miichaelsmedley** [comment]
I have seen projects where an eager product owner looking for features to be delivered would really love the VCA - but the decision tree makes the choice easy to understand.

**Comment by bahjatalaadelssw** [comment]
It's about scalability, and like any construction, it has it's limits.

**Reply by Calinator444** [reply]
In my opinion choosing the right software architecture is all about where you decide to separate concerns. If your data layer is susceptible to change then Clean Architecture is the best because database transactions are abstracted away into interfaces. If you want to easily add new features without a large blast radius than Vertical Slice architecture is generally the best fit. 

Granted if attending the  Software Architecture Whiterobes retreat taught me anything it's that there's usually more complexity in these patterns than it seems. Someone correct me if I'm wrong.

**Comment by awaltos** [comment]
CA seems easier to follow especially for people joining a large project. Either way it is important to think about this in advance and balance planning architecture whilst not locking into something that doesnt scale or just introduces complexity for the sake of it. Refactoring can be painful but sometimes that is needed early on in a project's life.

**Comment by ricksu978** [comment]
it gives clear logic and patterns, lets new team members ramp up fast, keeps code consistent, scales smoothly, and stops costly rewrites later.

**Comment by bettybondoc** [comment]
If possible, it’d be great to have templates for all four architectures listed here.

**Comment by kulesy** [comment]
Being able to foresee an application's growth and choosing the right architecture can help with scalability. I also agree with @bradystroud that sticking to an architecture is as (if not more) beneficial.

**Comment by amankumarrr** [comment]
I really liked the Clean Architecture approach; however, as rightly pointed out, even a simple change often required modifying multiple files. Vertical Slice Architecture addresses this pain point effectively.

**Comment by sethdaily** [comment]
This makes it understandable for those of us who are not developers - thanks

**Comment by AntPolkanov** [comment]
Choosing the right architecture at the start of a project saves heaps of time and money down the line. On my current project we are using MoMo, and it's been a huge win!

**Comment by PennyWalker** [comment]
I'm not a dev, but I know from an account management perspective well thought out, well designed project always save the client time, and money and often helps future proofs their project. Having a template would save so much time, and picking the right architecture from the beginning!

**Comment by jerryluossw** [comment]
> What do you think about the decision tree?

Great, it's easy to understand and can help us make quick decisions.

> Do you think it would be helpful when starting a new project?

Sure! This will save us a lot of time.

**Comment by wicksipedia** [comment]
“The right architecture” is a fallacy — there’s only “right for the time.”

Architecture must evolve with the product, team, and cost/benefit trade-offs. Over-investing early often backfires.

What matters most is that the code is malleable, well-encapsulated, and consistently structured so it can evolve without pain.

War story: I’ve seen over-engineered systems thrown out when real-world usage didn’t match assumptions. Start simple, adapt fast.

**Comment by suiyangqiu** [comment]
War Story: Starting a project, without getting your architecture right, it leads to unholy amounts of need for revision. In a few academic projects, I am ashamed to admit that I have dove straight into programming without getting the architecture right from the start of the project, resulting in huge refactoring being necessary later on that just sucked up time due to issues in the initial architecture. 😅

**Comment by ZenoWang1999** [comment]
Choosing an appropriate architecture is beneficial for the long-term maintenance of a project. More importantly, selecting a proper architecture means you have a clear understanding of what you want to accomplish and how your project should evolve.

**Comment by tino-liu** [comment]
> What do you think about the decision tree?

I think it's a pretty straightforward decision tree. Based on my current project, it either fits Clean Architecture or Vertical Slice

> Do you think it would be helpful when starting a new project?

Oh yeah

**Comment by sam-s-maher** [comment]
Working on a recent external project that despite it's size and complexity, had genuinely great architecture was quite a relief! It made onboarding and creating new functionality far smoother.

**Comment by thom-wang** [comment]
I am more of a front-end guy, so as you guys know, there is always a war between different front-end frameworks, especially between React and Vue. We argued a lot, and then we chose React because of the trust in Facebook, but we wasted a lot of time on the decision.

**Comment by SamBWagner** [comment]
On most of my projects, we try to adhere closely to clean architecture. However, with numerous developers working in isolated silos, it quickly becomes a disorganised situation. Architecture, code standards, and principles are only as effective as the people implementing them. People > Process

**Comment by jeoffreyfischer** [comment]
I've never been involved in a project from scratch, so the architecture was already set. 
But I've heard some stories of projects that started as a micro-service architecture and ended up like a massive spaghetti plate 🍝

**Comment by babakamyljanovssw** [comment]
The decision tree is helpful and based on that the Vertical Slice Architecture is more suitable for my current project.

**Comment by vladislav-kir** [comment]
I have not yet been the person responsible for choosing the architecture for the project, but mostly worked with Clean Architecture before. I do believe that ability of developers to adapt to changes is more critical than preaching a particular architecture.
